src/net_processing.cpp

```diff
@@ -5279,22 +5310,31 @@ void PeerManagerImpl::CheckForStaleTipAndEvictPeers()
 
     EvictExtraOutboundPeers(now);
 
+    Staleness staleness{Staleness::NotStale};
     if (now > m_stale_tip_check_time) {
-        // Check whether our tip is stale, and if so, allow using an extra
-        // outbound peer
-        if (!m_chainman.m_blockman.LoadingBlocks() && m_connman.GetNetworkActive() && m_connman.GetUseAddrmanOutgoing() && TipMayBeStale()) {
+        // Check whether our tip is stale, and if so, allow using an extra outbound peer
+        if (!m_chainman.m_blockman.LoadingBlocks() && m_connman.GetNetworkActive() && m_connman.GetUseAddrmanOutgoing()) {
+            staleness = TipMayBeStale();
+        }
+
+        if (staleness != Staleness::NotStale) {
             LogPrintf("Potential stale tip detected, will try using extra outbound peer (last tip update: %d seconds ago)\n",
                       count_seconds(now - m_last_tip_update.load()));
             m_connman.SetTryNewOutboundPeer(true);
         } else if (m_connman.GetTryNewOutboundPeer()) {
             m_connman.SetTryNewOutboundPeer(false);
         }
+
         m_stale_tip_check_time = now + STALE_CHECK_INTERVAL;
     }
 
     if (!m_initial_sync_finished && CanDirectFetch()) {
         m_connman.StartExtraBlockRelayPeers();
         m_initial_sync_finished = true;
+    } else if (staleness == Staleness::StalePastNetworkLimited) {
+        // Disallow connection to limited peers.
+        // We need to find a full node that can help us recover.
+        m_initial_sync_finished = false;
```

Comment:
The word `initial` here is indeed confusing... Thinking about something happening only at node bootstrap, which is a more obvious version i think. Can we find a different word now that you're changing it to certainly not mean that?

Reply:
> The word `initial` here is indeed confusing... Thinking about something happening only at node bootstrap, which is a more obvious version i think. Can we find a different word now that you're changing it to certainly not mean that?

Hmm, first names coming to my mind are `m_close_to_synced` or `m_close_to_tip`? or, could use the existing term `m_can_direct_fetch`?

Still, maybe better to leave it for a follow-up because, with the back-and-forth functionality, we could also use this flag instead of calling `CanDirectFetch()` which locks `cs_main`.

Reply:
> This flag is only used to connect to limited peers or not. In that case, why not remove the indirection about syncing and just call it `allow_limited_peers`?

Because of what I wrote above, this flag can (and most likely will) be used in other places as well, fulfilling two goals: (1) minimizing `cs_main` locks, which improves the overall node response, and (2) minimizing the usage of the chainstate IBD flag, which lacks back-and-forth functionality

Reply:
Right, then I think `m_is_close_to_tip` would make the most sense, since it's set initially after calling `IsBlockCloseToTip`.

Reply:
@naumenkogs, do you agree on the naming as well? I'm happy to change it if we all agree.
Better to be sync to not circle much around the naming here.

Reply:
yeah that would be better i think

Reply:
ok cool. Updated per feedback.